File: Binary Search/minimumEffortPath.js
const minimumEffortPath = (heights) => {
  const check = (effort) => {
    const directions = [
      [-1, 0],
      [1, 0],
      [0, -1],
      [0, 1],
    ];

    const m = heights.length;
    const n = heights[0].length;

    const isValid = (row, col) => row >= 0 && row < m && col >= 0 && col < n;
    const isAcceptableEffort = (prevNum, nextNum) =>
      Math.abs(prevNum - nextNum) <= effort;

    const seen = [];
    for (let row = 0; row < m; row++) seen.push(new Array(n).fill(false));
    seen[0][0] = true;

    const stack = [[0, 0]];

    while (stack.length) {
      let [row, col] = stack.pop();
      for (let [x, y] of directions) {
        let newRow = row + y;
        let newCol = col + x;
        if (
          isValid(newRow, newCol) &&
          !seen[newRow][newCol] &&
          isAcceptableEffort(heights[row][col], heights[newRow][newCol])
        ) {
          if (newRow === m - 1 && newCol === n - 1) return true;
          seen[newRow][newCol] = true;
          stack.push([newRow, newCol]);
        }
      }
    }

    return false;
  };

  let left = 0,
    right = 0;
  for (let row of heights) {
    for (let num of row) right = Math.max(right, num);
  }

  while (left <= right) {
    const mid = Math.floor((left + right) / 2);
    if (check(mid)) {
      right = mid - 1;
    } else {
      left = mid + 1;
    }
  }

  return left;
};

// Time complexity O(mnlogk)
// Space O(mn)

const heights = [
  [1, 2, 2],
  [3, 8, 2],
  [5, 3, 5],
];

console.log(minimumEffortPath(heights));

File: Binary Search/MinimumEffortPath.java
import java.util.ArrayDeque;
import java.util.Arrays;

public class MinimumEffortPath {
  int[][] heights;
  int m;
  int n;

  static class HeightsPosition {
    int row;
    int col;

    public HeightsPosition(int row, int col) {
      this.row = row;
      this.col = col;
    }
  }

  public int minimumEffortPath(int[][] heights) {
    this.heights = heights;
    this.m = heights.length;
    this.n = heights[0].length;

    int left = 0;
    int right = 0;
    for (int[] row : heights) {
      for (int num : row)
        right = Math.max(right, num);
    }

    while (left <= right) {
      int mid = left + (right - left) / 2;
      if (check(mid)) {
        right = mid - 1;
      } else {
        left = mid + 1;
      }
    }

    return left;
  }
  
  public boolean check(int effort) {
    int[][] directions = new int[][] { { -1, 0 }, { 1, 0 }, { 0, -1 }, { 0, 1 } };
    boolean[][] seen = new boolean[m][n];
    seen[0][0] = true;

    ArrayDeque<HeightsPosition> stack = new ArrayDeque<>();
    stack.push(new HeightsPosition(0, 0));

    while (!stack.isEmpty()) {
      HeightsPosition position = stack.removeLast();
      for (int[] direction : directions) {
        int x = direction[0];
        int y = direction[1];
        HeightsPosition newPosition = new HeightsPosition(position.row + y, position.col + x);
        if (isValid(newPosition.row, newPosition.col) && !seen[newPosition.row][newPosition.col]
            && isAcceptableEffort(position, newPosition, effort)) {
          if (newPosition.row == m - 1 && newPosition.col == n - 1)
            return true;
          seen[newPosition.row][newPosition.col] = true;
          stack.addLast(newPosition);
        }
      }
    }

    return false;
  }
  
  public boolean isValid(int row, int col) {
    return row >= 0 && row < m && col >= 0 && col < n;
  }

  public boolean isAcceptableEffort(HeightsPosition lastPosition, HeightsPosition newPosition, int allowableEffort) {
    return Math.abs(heights[lastPosition.row][lastPosition.col] - heights[newPosition.row][newPosition.col]) <= allowableEffort;
  }
}